### PR TITLE
Rails 7

### PR DESCRIPTION
**Pull request by mcbloch**
Big update. Still work in progress.
Ruby logic is transfered and config is almost right except capistrano.
Assets pipeline and scripts are being worked on.

Todo:
- [x] Refresh CSS library, perhaps Tailwind?
   - [x] Restyle global UI
   - [x] Restyle sign in page
   - [x] Restyle landing page
   - [x] Restyle transaction form
   - [x] Restyle transaction page
   - [x] Restyle requests page
   - [x] Restyle notification page
   - [x] Add footer
- [x] Big javascript cleanup:
   - [x] Fix/find alternative/remove chart library (and specifically chart on the home page) -> eh, won't do, effort with not that much gain and it's even privacy invasive!
   - [x] Fix/find alternative/remove select2 library -> Went for React-Select (in React components) and TomSelect in non-react components. TomSelect was chosen as it's "pure" library without jquery/other framework dependencies
   - [x] Fix/find alternative/remove datatables library (and tables on the request page)
   - [x] ~no more jquery~ -> can't do, datatables depends on this, could be replaced later on
   - [x] clean up Transactionform
   - [x] I pulled in`remount` (see https://github.com/ZeusWPI/Tab/commit/1c24ee5a723fb91b122e94a5f13dbf17137c07ee#diff-9ce245a7494222fb60b9bc259db7275af276282f0e91eb7bd3af02b048f60b81), but this might be overkill, killed it and embraced Stimulus completely
   - [x] Introduce a HMR library -> yeah, nah
   - [x] Add errors and confirmation back in transfer form on home page
   - [x] Add errors and confirmation back in transaction form
   - [x] Add flash back
   - [x] Add logout button back
- [x] Fix interesting readme elements removed in this PR 
- [x] Check mobile support
- [x] Omniauth 2 and higher

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,16 +1,24 @@
...
 
-## Adding clients
+* Ruby version
 
-A client can see and modify balances of other users.
+* System dependencies
 
-To add a client, connect
-to the server, `cd production/current`, then run `RAILS_ENV=production bundle exec rails console`.
-Then you can add clients with `client = Client.create name: "Tap"`.
```
Comment (redfast00) [inline README.md:14]: Dit is op zich wel nog nuttige info

File: Gemfile [hunk 2]
```diff
@@ -1,106 +1,110 @@
...
+gem 'turbo-rails'
+
+# Hotwire's modest JavaScript framework [https://stimulus.hotwired.dev]
+gem 'stimulus-rails'
+
+# Build JSON APIs with ease [https://github.com/rails/jbuilder]
+gem 'jbuilder'
+
+# Use Redis adapter to run Action Cable in production
+gem 'redis', '~> 4.0'
```
Comment (redfast00) [inline Gemfile:31]: Gaan we Redis in production nodig hebben?

File: Gemfile [hunk 3]
```diff
@@ -1,106 +1,110 @@
...
+# gem "image_processing", "~> 1.2"
 
 # Authentication
 gem 'devise'
-gem 'omniauth-oauth2', '~> 1.3.1'
+gem 'omniauth-oauth2', '~> 1.3.1' # '~>1.7'
+gem 'omniauth-rails_csrf_protection'
 
 # Authorisation
 gem 'cancancan'
```
Comment (redfast00) [inline Gemfile:57]: Feel free om hier iets anders voor te gebruiken als je dat leuker lijkt, ik heb opgevangen dat cancan mss niet supernice is hiervoor

Comment (TomNaessens) [reply Gemfile:54]: Is wel OK, pundit kan ook maar dan buiten deze PR. Cancancan is de onderhouden versie van can & cancan.

File: README.md (hunk 1 above)
Comment (TomNaessens) [reply README.md:14]: Readded.

File: Gemfile (hunk 2 above)
Comment (TomNaessens) [reply Gemfile:31]: Nee, we hebben geen jobs.

File: app/javascript/controllers/select-controller.js [hunk 4]
```diff
@@ -0,0 +1,11 @@
+import { Controller } from "@hotwired/stimulus";
+
+import TomSelect from "tom-select";
+
+export default class extends Controller {
+  connect() {
+    new TomSelect("#transaction_creditor", {
```
Comment (redfast00) [inline app/javascript/controllers/select-controller.js:7]: Zorgt dit ervoor dat je enkel Tom kan selecteren als creditor? Want daar ben ik misschien niet helemaal akkoord mee...

Comment (TomNaessens) [reply app/javascript/controllers/select-controller.js:7]: Nee, helaas. TomSelect is een select2/typehead/... alternatief dat geen framework dependency (jquery, react, vue, ...) heeft.

File: .github/workflows/ruby.yml [hunk 5]
```diff
@@ -11,18 +11,18 @@ jobs:
...
+
     - name: Set up Ruby
       uses: ruby/setup-ruby@v1
       with:
-        ruby-version: 2.5
-  
+        ruby-version: "3.1.2"
+
     - name: Install bundler
       run: gem install bundler -v "$(grep -A 1 BUNDLED.WITH Gemfile.lock | tail -n 1)"
```
[the remark this answers is not in the document]
Comment (TomNaessens) [reply .github/workflows/ruby.yml]: Nice!

File: .github/workflows/ruby.yml [hunk 6]
```diff
@@ -11,18 +11,18 @@ jobs:
...
     steps:
     - uses: actions/checkout@v2
-    
+
     - name: Set up Ruby
       uses: ruby/setup-ruby@v1
       with:
-        ruby-version: 2.5
-  
+        ruby-version: "3.1.2"
```
[the remark this answers is not in the document]
Comment (TomNaessens) [reply .github/workflows/ruby.yml]: Nice!

**Comment by mcbloch** [comment]
When trying to log in I get the following error from adams: 
```
Not acceptable

Sorry, but you're not allowed to do this.
```
The url I'm redirected to is: https://adams.ugent.be/oauth/authorize/?client_id&redirect_uri=http%3A%2F%2Flocalhost%3A3000%2Fusers%2Fauth%2Fzeuswpi%2Fcallback&response_type=code&state=1678985c91e81372568b51874ed29341ed0f68267a4f7f5f

I think there should be a client_id and client_secret in there.
If I look at the development credentials, I don't see the `omniauth_client_{id, secret}` mentioned in `config/initializers/devise.rb`. Have you set these on your machine?

As credentials are encrypted they can safely be commited into git.

----

On another note:

Latest nodejs 16 version is `16.15.0`. We pinned `16.5.0` for this project. Can we upgrade to the latest 16 version? `17` and `18` are also out, we could potentially already upgrade to those?

**Review by redfast00 — APPROVED** [review]

**Pull request merged**